src/body-builder.js

```diff
@@ -0,0 +1,197 @@
+    Object.assign(clause, opts, {[field]: value})
+  } else if (field && value) {
+    clause[field] = value
+  } else if (field) {
+    clause.field = field
+  }
+
+  return clause
+}
+
+function queryBuilder () {
+  let query = {}
+
+  function makeQuery (boolType, queryType, ...args) {
+    const nested = {}
+    if (_.isFunction(_.last(args))) {
+      const nestedCallback = args.pop()
+      const nestedResult = nestedCallback(
+        Object.assign(queryBuilder(), filterBuilder())
+      )
+      if (_.size(nestedResult.getQuery())) {
+        nested.query = nestedResult.getQuery()
+      }
+      if (_.size(nestedResult.getFilter())) {
+        nested.filter = nestedResult.getFilter()
+      }
+    }
+
+    query = boolMerge(
+      {[queryType]: Object.assign(buildClause(...args), nested)},
+      query,
+      boolType
+    )
+  }
+
+  return {
+    query (...args) {
+      makeQuery('and', ...args)
+      return this
+    },
+    andQuery (...args) {
+      return this.query(...args)
+    },
+    addQuery (...args) {
+      return this.query(...args)
+    },
+    orQuery (...args) {
+      makeQuery('or', ...args)
+      return this
+    },
+    notQuery (...args) {
+      makeQuery('not', ...args)
+      return this
+    },
+    getQuery () {
+      return query
+    }
+  }
+}
+
+function filterBuilder () {
+  let filter = {}
+
+  function makeFilter (boolType, filterType, ...args) {
+    const nested = {}
+    if (_.isFunction(_.last(args))) {
+      const nestedCallback = args.pop()
+      const nestedResult = nestedCallback(
+        Object.assign(queryBuilder(), filterBuilder(), aggregationBuilder())
+      )
+      if (_.size(nestedResult.getQuery())) {
+        nested.query = nestedResult.getQuery()
+      }
+      if (_.size(nestedResult.getFilter())) {
+        nested.filter = nestedResult.getFilter()
+      }
+      if (_.size(nestedResult.getAggregations())) {
+        nested.aggs = nestedResult.getAggregations()
+      }
+    }
+
+    filter = boolMerge(
+      {[filterType]: Object.assign(buildClause(...args), nested)},
+      filter,
+      boolType
+    )
+  }
+
+  return {
+    filter (...args) {
+      makeFilter('and', ...args)
+      return this
+    },
+    andFilter (...args) {
+      return this.filter(...args)
+    },
+    addFilter (...args) {
+      return this.filter(...args)
+    },
+    orFilter (...args) {
+      makeFilter('or', ...args)
+      return this
+    },
+    notFilter (...args) {
+      makeFilter('not', ...args)
+      return this
+    },
+    getFilter () {
+      return filter
+    }
+  }
+}
+
+function buildAggregation (type, field, opts) {
+  return {
+    [type]: Object.assign({field}, opts)
+  }
+}
+
+function aggregationBuilder () {
+  const aggregations = {}
+
+  function makeAggregation (type, field, ...args) {
+    const aggName = _.find(args, _.isString) || `agg_${type}_${field}`
+    const opts = _.find(args, _.isPlainObject) || {}
+    const nested = _.find(args, _.isFunction)
```

Comment:
I'm actually slightly unhappy with it, but saw no better way here. Reason is as you stated, that it doesn't work anymore as soon as you have two arguments of the same type.

All in all I'm rather unhappy with callback-style nesting. I'd much prefer if the user could use it like
```js
const postAggregation = aggregationBuilder().aggregation('terms', 'postId')
const body = bodyBuilder()
  .aggregation('terms', 'user', postAggregation)
  .build()
```

in this case I would get all the users that posted comments and all the post IDs for which they left comments... I find it a nicer way of nesting.

Would be happy to get your input on this :)